### PR TITLE
Feature kvm ifdown script

### PR DESCRIPTION
**Pull request by yiannist**
This patch-set adds support for an ifdown script that is responsible for deconfiguring network devices and cleanup changes made by the ifup script as described in the corresponding [design-doc](http://docs.ganeti.org/ganeti/2.15/html/design-ifdown.html). While at it, we change the `kvm-if-bridge` to `kvm-ifup-custom` (for consistency) and we pass positional arguments to those custom scripts. We use this functionality for many years in GRNET but I just found some time to cleanup the code and share. The original work was done by @dimara.

### CODE REVIEW

**Comment by dimara** [comment]
Hi, I am totally fine with the PR. This is a old work that finally got submitted! Many thanks to @yiannist!

PS: Here is also a [relevant thread](https://groups.google.com/forum/#!msg/ganeti-devel/5-duINuErTk/auGojMPaAgAJ) that could be useful during review.

**Comment by dimara** [comment]
Maybe, to keep backwards compatibility, the kvm-ifup.in file should still exec the deprecated kvm-vif-bridge script (if found).

**Comment by yiannist** [comment]
Yeap, that's a good point! I just updated the relevant commit.

**Comment by iustin** [comment]
Friendly ping on this.

File: lib/hypervisor/hv_base.py [hunk 1]
```diff
@@ -209,30 +209,32 @@ def GenerateTapName():
...
+  @param nic: The NIC object
+  @type tap: string
+  @param tap: The tap name
   @type seq: int
-  @param seq: nic sequence number
-  @type nic: nic object
-  @param nic: nic we're acting on
-  @type tap: str
-  @param tap: the host's tap interface this NIC corresponds to
+  @param seq: The NIC's sequense number
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_base.py]: ACK.

File: lib/hypervisor/hv_base.py [hunk 2]
```diff
@@ -251,9 +253,39 @@ def ConfigureNIC(cmd, instance, seq, nic, tap):
...
+  @param instance: instance we're acting on
+  @type seq: int
+  @param seq: nic sequence number
+  @type nic: nic object
+  @param nic: nic we're acting on
+  @type tap: str
+  @param tap: the host's tap interface this NIC corresponds to
+
+  """
+  env = CreateNICEnv(instance.name, nic, tap, seq, instance.GetTags())
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_base.py]: ACK.

File: lib/hypervisor/hv_kvm/__init__.py [hunk 3]
```diff
@@ -802,11 +802,40 @@ def _InstanceNICDir(cls, instance_name):
...
+    """Returns the tap for the corresponding nic
+
+    Search for tap file named after NIC's uuid.
+    For old instances without uuid indexed tap files returns nothing.
+
+    """
+    try:
+      return utils.ReadFile(cls._InstanceNICFile(instance_name, nic.uuid))
+    except EnvironmentError:
+      pass
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: ACK.

File: lib/hypervisor/hv_kvm/__init__.py [hunk 4]
```diff
@@ -802,11 +802,40 @@ def _InstanceNICDir(cls, instance_name):
...
+
+  @classmethod
+  def _WriteInstanceNICFiles(cls, instance_name, seq, nic, tap):
+    """Write tap name to both instance NIC files
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: ACK

File: lib/hypervisor/hv_kvm/__init__.py [hunk 5]
```diff
@@ -802,11 +802,40 @@ def _InstanceNICDir(cls, instance_name):
...
+
+    """
+    for ident in [seq, nic.uuid]:
+      utils.WriteFile(cls._InstanceNICFile(instance_name, ident), data=tap)
+
+  @classmethod
+  def _RemoveInstanceNICFiles(cls, instance_name, seq, nic):
+    """Write tap name to both instance NIC files
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: ACK.

File: lib/hypervisor/hv_kvm/__init__.py [hunk 6]
```diff
@@ -888,6 +917,29 @@ def _ConfigureNIC(instance, seq, nic, tap):
     """
     hv_base.ConfigureNIC([pathutils.KVM_IFUP, tap], instance, seq, nic, tap)
 
+  @classmethod
+  def _UnconfigureNIC(cls, instance_name, nic, only_local=True):
+    """Run ifdown script for a specific NIC
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: ACK.

File: lib/hypervisor/hv_base.py [hunk 7]
```diff
@@ -251,9 +253,39 @@ def ConfigureNIC(cmd, instance, seq, nic, tap):
     n = objects.Network.FromDict(nic.netinfo)
     env.update(n.HooksDict())
 
+  if instance_tags:
+    env["TAGS"] = " ".join(instance_tags)
+
+  # This should always be available except for old instances in the
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_base.py]: Not sure what we should do in that case. I suppose that we could add a "special" value to indicate that we have an _old_ instance and let the hook decide? @dimara, what do you think about this?

The ifdown [design doc](http://docs.ganeti.org/ganeti/2.15/html/design-ifdown.html#implementation-details) states for that:
> Obviously this info will not be available for old instances in the cluster.

File: lib/hypervisor/hv_kvm/__init__.py [hunk 8]
```diff
@@ -917,6 +917,29 @@ def _ConfigureNIC(instance, seq, nic, tap):
...
+    @type instance_name: string
+    @param nic: nic we're acting on
+    @type nic: nic object
+    @param only_local: whether ifdown script should reset global conf (dns)
+    @type only_local: boolean
+
+    """
+    tap = cls._GetInstanceNICTap(instance_name, nic)
+    env = hv_base.CreateNICEnv(instance_name, nic, tap)
+    arg2 = 'true' if only_local else 'false'
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: It is true that there is an asymmetry here. However, I don't see a use-case for having something similar on the ifup script. I think that this only makes sense for the ifdown functionality as described also in doc. Could you explain more the scenario that you might have the configuration executed **twice** on during migration? I think that the configuration will happen only once on the new node. @dimara, feel free to correct me, if I am missing something.

File: lib/hypervisor/hv_kvm/__init__.py [hunk 9]
```diff
@@ -2420,6 +2421,7 @@ def CleanupInstance(self, instance_name):
     pidfile, pid, alive = self._InstancePidAlive(instance_name)
     if pid > 0 and alive:
       raise errors.HypervisorError("Cannot cleanup a live instance")
+    self._UnconfigureInstanceNICs(instance_name)
```
[the remark this answers is not in the document]
Comment (yiannist) [reply lib/hypervisor/hv_kvm/__init__.py]: Based on:
> Additionally it should modify only the node local dynamic configs (routes, arp entries, SDN, firewalls, etc.), whereas static ones (DNS, DHCP, etc.) should be modified via hooks.

I would say that this is the intended behavior but I will also let @dimara correct me! :)

**Comment by rbott** [comment]
I would propose to close this PR and give it a shot against the current master (and address the open questions here). I will open a feature request issue to track it. I might open a PR based on this here myself, but this might take a couple of weeks. Are you fine with that @yiannist ?

**Comment by rbott** [comment]
I'll move forward with this and close the PR for now. If it is still required, it has to be re-opened against the current master branch :-(

**Pull request closed without merging**